### PR TITLE
expression, parser: support GBK for builtin function ord

### LINKED ISSUE #29463: Support GBK for builtin function `ord`

### PR DESCRIPTION
**Pull request by unconsolable**
Signed-off-by: unconsolable <chenzhipeng2012@gmail.com>

### What problem does this PR solve?

Issue Number: close #29463 

Problem Summary:

### What is changed and how it works?
- Add `(*Encoding).EncodeFirstChar` to get first utf8 code point and encode.
- Modify `chooseOrdFunc`, `ordSingleByte`, `ordOthers` to support GBK
- Add unit test & explain test

### Check List

Tests

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by zimulala** [comment]
@unconsolable Please fix this conflict.

**Comment by unconsolable** [comment]
> @unconsolable Please fix this conflict.

Resolved.

File: expression/builtin_string.go [hunk 1]
```diff
@@ -2916,45 +2916,53 @@ func (b *builtinOrdSig) Clone() builtinFunc {
 // See https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_ord
 func (b *builtinOrdSig) evalInt(row chunk.Row) (int64, bool, error) {
 	str, isNull, err := b.args[0].EvalString(b.ctx, row)
+	strBytes := hack.Slice(str)
```
Comment (zimulala) [inline expression/builtin_string.go:2919]: I think putting it before https://github.com/pingcap/tidb/pull/29527/files#diff-159c25a0ed92223cdf33859a7b1d5d25d7d018136ba44c4eabdac0d1619d2138R2931 is better. We can avoid unnecessary calls in some cases

File: expression/builtin_string.go [hunk 2]
```diff
@@ -2916,45 +2916,53 @@ func (b *builtinOrdSig) Clone() builtinFunc {
...
-func chooseOrdFunc(charSet string) (func(string) int64, error) {
+func chooseOrdFunc(charSet string) (func([]byte) int64, error) {
 	// use utf8 by default
 	if charSet == "" {
 		charSet = charset.CharsetUTF8
 	}
 	desc, err := charset.GetCharsetInfo(charSet)
 	if err != nil {
-		return nil, err
+		// TODO: After gbk added to charset, this compare should be removed
```
Comment (zimulala) [inline expression/builtin_string.go:2945]: Could we remove this and put it to `TestOrd`(comment out related to the test )? I think we can pass tests in `new_character_set_builtin.test`.

**Comment by zimulala** [comment]
PTAL @xiongjiwei @tangenta

File: expression/builtin_string.go (hunk 1 above)
Comment (unconsolable) [reply expression/builtin_string.go]: Resolved.

File: expression/builtin_string.go (hunk 2 above)
Comment (unconsolable) [reply expression/builtin_string.go]: Resolved.

**Review by zimulala — APPROVED** [review]
LGTM

File: expression/builtin_string_test.go [hunk 3]
```diff
@@ -2226,21 +2226,27 @@ func TestOrd(t *testing.T) {
...
+		{23, 50, "", false, false},
+		{2.3, 50, "", false, false},
+		{nil, 0, "", true, false},
+		{"", 0, "", false, false},
+		{"你好", 14990752, "", false, false},
+		{"にほん", 14909867, "", false, false},
+		{"한국", 15570332, "", false, false},
+		{"👍", 4036989325, "", false, false},
+		{"א", 55184, "", false, false},
+		// TODO: Uncomment it when gbk be added into charsetInfos
```
Comment (tangenta) [inline expression/builtin_string_test.go:2245]: Please use `SetCharsetFeatEnabledForTest` to enable the new-character-set feature.

Comment (unconsolable) [reply expression/builtin_string_test.go]: Resolved.

**Comment by zimulala** [comment]
PTAL @tangenta

**Review by tangenta — APPROVED** [review]

**Pull request merged**